README
This branch is closed. Further development will happen at http://git.sugarlabs.org/surf.


NEWS
86

* Translations release

85

* Remove canceled download from active downloads list #6018 (erikos)

84

* Use ellipsis, #5765 (rwh)
* Implement can_close(), #5493 (rwh)

82

* Enable gobject threads, fix #5639 (marco)

81

* Use the new show_object_in_journal API of the base activity class
  to show a downladed object or the view source one in the journal
  #4909 (cscott)

80

* Update the progress of downloads less often #5449 (tomeu)

79
* Missing file for: Open an uri of type text/uri-list with browse #5080 
(erikos)

78

* Give to the DS the ownership of the files we hand to it #5235 (tomeu)
* Open an uri of type text/uri-list with browse #5080 (erikos)
* Save view source to SAR/instance #5251 (erikos)
* Move the profile in SAR/data #5221 (erikos)

77

76

* Added OLPC root CA support (marco, erikos)

75

* fix for fn+space (erikos)
* #4909 do not resume source with write (erikos)
* #4909 do not resume downloaded files (erikos)

74

* change the LIBRARY_PATH to point to 
  /usr/share/library-common/index.html (erikos)

73

* #4909: added a dbus call to open the source data by the journal (marco)
* #4889: use the window's fullscreen method to get to fullscreen mode,
this does not mess up the state the window manager has (erikos)
* #4870: Add onStatusChange method to Download. (tomeu)
* #4904: Save downloads to the instance dir, not tmp. (tomeu)

72

* Save the page source with mime type text/html (erikos)

71

* Added zoom support with keybindings (erikos)
* Added a view toolbar for fullscreen/tray/zoom (erikos)
* Make outline of the session bookmarks smaller (erikos)
* Use favorite icon for session bookmark and call it Bookmark #3741 (erikos)

70

* Moved the handling from hide/show of the tray to activity window (erikos)

69

* Remove download object when download is canceled when downloaded already 
as well (erikos)
* Make use of the tray support in sugar activity window (erikos)

68

* Adopt 'go' icons (marco)

67

* Use ctrl+t for tray visibility, this should move to sugar-graphics-window (erikos)
* Fix the case where no activity can be associated with the downloaded file (erikos)
* Changed icon for stop button, place alert on top (erikos)

66

* Adapt to sugar API change (marco)

65

* Removed hack to set the right service name for sharing (erikos)
* Removed the clipboard part from the download manager and added alerts (erikos)

64

63

* Fix crash on startup. (marco)
* #3552: Cancel a donwload when the related journal entry is deleted. (tomeu)

62

* #2910 Open popups in the current window (marco)
* Add agent-stylesheet.css and support for blocking flash. (tomeu)
* Add support for agent and user stylesheets. (tomeu)

61

* Adopted to the latest actviity launch changes (erikos)
* Took the import of securitydialog out (erikos)

60

* Limit to 15 entries in the palette of the back/forward buttons (erikos)
* Always start at the first toolbar page (erikos)

59

* Add several ogg subtypes (marco)

57

* Fix a typo, make mime types actually work (marco)

56

* Add more ogg mime types (marco)

55

* Backward compatibility of journal entries to trial-2 (erikos)

54

* Update spanish translation.

53

* Make hide/show of the treay one button #3475 (erikos)

52

* Fix for when participant leaves (erikos)
* #3347 Advertise ogg and wav mime types (marco)
* Added the buttons for hide/show tray (erikos)

51

* #3339 Add some mime types web-activity can handle (marco)
* Added support for the view source key fn+space (erikos)

50

* View source for Browse
  With the ctrl+u command does load the current page from the cache and 
  launches write to display it #3261 (erikos)
* Changed the show/hide frame key to be ctrl+v for visibility (erikos)
  
49

* Added missing fill_color in icon (erikos)
* #2555: Report download progress to the Journal via the DataStore. (tomeu)

48

* New activity icon, Fix for #2829 (erikos)

47

* redesign when members leave and rejoin, added timestamp to be 
able to bring the links in the right order, you can not add alink 
twice unless you deleted it already, links are added to the tray 
on the right side (erikos)

46
* fix for gray page when resuming from journal (erikos)
* #2327 Unescape the URI to display it in the address bar (marco)
* fixed memory consuption when creating shared links 
  by calling the garbage collector (erikos)

45

* use json for storing the history and the shared
  link information to the journal (erikos)
* use sugar tray which enables scrolling for the links (erikos)
* only try to connect to presence service when offline (erikos)

44

* new design for shared links (erikos)
* start the activity with in the browser tab open when in joined 
   a shared session or a handle.uri is given (erikos)

43

* fix for Palette.TOP (erikos)

42

* Added the possibility to hide and show the linktray. currently you can toggle this with ctrl+s(erikos)
* Port to new tubes API. (cassidy)
* Use sugar's tubeconn.py. (cassidy)

41

* Code cleanup. (simon)

40

* #2304 This adds the sharing facilities to the web-activity. 

You share the currently loaded page by hitting the star button in the browser toolbbar when you are in the shared mode of the activity. The link will come up as a thumbnail in the toolbar at the bottom of the page. The information about the link: url, title, thumbnail, creator gets then transmitted to all the other members. They can then decide whether to show the link in their browser or not. Everyone can add a link but you can only delete a link in your local linkbar which prevents someone from erasing all your links. (erikos)

* #2577: Add german translation. (fab)
* #2596: Add french translation. (Samuel Bizien)

38

* #2763: Save downloads to flash, not to /tmp. (tomeu)

37

* Add greek translation. (simosx)

36

* Increase timeout when copying the file into the datastore to 360s. (tomeu)

35

* #2370 Updated spanish translation. (xavi)

34

* #2364: Fix uploads and clean up correctly the temp files. (tomeu)

33


webactivity.py
# Copyright (C) 2006, Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import os
import logging
from gettext import gettext as _

import gtk
import dbus
import sha
import base64
import time
import shutil
 
from sugar.activity import activity
from sugar import env
from sugar.graphics import style
import telepathy
import telepathy.client
from sugar.presence import presenceservice
from sugar.graphics.tray import HTray
from sugar import profile
from sugar.graphics.alert import Alert
from sugar.graphics.icon import Icon

PROFILE_VERSION = 1

_profile_version = 0
_profile_path = os.path.join(env.get_profile_path(), 'gecko')
_version_file = os.path.join(_profile_path, 'version')

if os.path.exists(_version_file):
    f = open(_version_file)
    _profile_version = int(f.read())
    f.close()

if _profile_version < PROFILE_VERSION:
    if not os.path.exists(_profile_path):
        os.mkdir(_profile_path)

    shutil.copy('cert8.db', _profile_path)
    os.chmod(os.path.join(_profile_path, 'cert8.db'), 0600)

    f = open(_version_file, 'w')
    f.write(str(PROFILE_VERSION))
    f.close()

import hulahop
hulahop.startup(_profile_path)

from browser import Browser
from webtoolbar import WebToolbar
from viewtoolbar import ViewToolbar
import downloadmanager
import promptservice
import filepicker
import sessionhistory 
import progresslistener

_LIBRARY_PATH = '/usr/share/library-common/index.html'

from model import Model
from sugar.presence.tubeconn import TubeConnection
from messenger import Messenger
from linkbutton import LinkButton

SERVICE = "org.laptop.WebActivity"
IFACE = SERVICE
PATH = "/org/laptop/WebActivity"

_VIEW_SOURCE_KEY_CODE = 0x1008FF1A
_TOOLBAR_BROWSE = 1

_logger = logging.getLogger('web-activity')

class WebActivity(activity.Activity):
    def __init__(self, handle):
        activity.Activity.__init__(self, handle)

        _logger.debug('Starting the web activity')

        self._browser = Browser()

        temp_path = os.path.join(self.get_activity_root(), 'instance')
        downloadmanager.init(self._browser, self, temp_path)
        sessionhistory.init(self._browser)
        progresslistener.init(self._browser)

        toolbox = activity.ActivityToolbox(self)
        activity_toolbar = toolbox.get_activity_toolbar()

        self.toolbar = WebToolbar(self._browser)
        toolbox.add_toolbar(_('Browse'), self.toolbar)
        self.toolbar.show()
       
        self._tray = HTray()
        self.set_tray(self._tray, gtk.POS_BOTTOM)
        self._tray.show()
        
        self.viewtoolbar = ViewToolbar(self)
        toolbox.add_toolbar(_('View'), self.viewtoolbar)
        self.viewtoolbar.show()

        self.set_toolbox(toolbox)
        toolbox.show()

        self.set_canvas(self._browser)
        self._browser.show()
                 
        self.session_history = sessionhistory.get_instance()
        self.session_history.connect('session-link-changed', self._session_history_changed_cb)
        self.toolbar.connect('add-link', self._link_add_button_cb)

        self._browser.connect("notify::title", self._title_changed_cb)

        self.model = Model()
        self.model.connect('add_link', self._add_link_model_cb)

        self.current = _('blank')
        self.webtitle = _('blank')
        self.connect('key-press-event', self._key_press_cb)
                     
        self.toolbox.set_current_toolbar(_TOOLBAR_BROWSE)
        
        if handle.uri:
            self._browser.load_uri(handle.uri)        
        elif not self._jobject.file_path:
            # TODO: we need this hack until we extend the activity API for
            # opening URIs and default docs.
            self._load_homepage()

        self.messenger = None
        self.connect('shared', self._shared_cb)

        # Get the Presence Service        
        self.pservice = presenceservice.get_instance()
        try:
            name, path = self.pservice.get_preferred_connection()
            self.tp_conn_name = name
            self.tp_conn_path = path
            self.conn = telepathy.client.Connection(name, path)
        except TypeError:
            _logger.debug('Offline')
        self.initiating = None
            
        if self._shared_activity is not None:
            _logger.debug('shared:  %s' %self._shared_activity.props.joined)

        if self._shared_activity is not None:
            # We are joining the activity
            _logger.debug('Joined activity')                      
            self.connect('joined', self._joined_cb)
            if self.get_shared():
                # We've already joined
                self._joined_cb()
        else:   
            _logger.debug('Created activity')
    
    def _shared_cb(self, activity):
        _logger.debug('My activity was shared')        
        self.initiating = True                        
        self._setup()

        _logger.debug('This is my activity: making a tube...')
        id = self.tubes_chan[telepathy.CHANNEL_TYPE_TUBES].OfferDBusTube(
                SERVICE, {})
                
    def _setup(self):
        if self._shared_activity is None:
            _logger.debug('Failed to share or join activity')
            return

        bus_name, conn_path, channel_paths = self._shared_activity.get_channels()

        # Work out what our room is called and whether we have Tubes already
        room = None
        tubes_chan = None
        text_chan = None
        for channel_path in channel_paths:
            channel = telepathy.client.Channel(bus_name, channel_path)
            htype, handle = channel.GetHandle()
            if htype == telepathy.HANDLE_TYPE_ROOM:
                _logger.debug('Found our room: it has handle#%d "%s"' 
                    %(handle, self.conn.InspectHandles(htype, [handle])[0]))
                room = handle
                ctype = channel.GetChannelType()
                if ctype == telepathy.CHANNEL_TYPE_TUBES:
                    _logger.debug('Found our Tubes channel at %s'%channel_path)
                    tubes_chan = channel
                elif ctype == telepathy.CHANNEL_TYPE_TEXT:
                    _logger.debug('Found our Text channel at %s'%channel_path)
                    text_chan = channel

        if room is None:
            _logger.debug("Presence service didn't create a room")
            return
        if text_chan is None:
            _logger.debug("Presence service didn't create a text channel")
            return

        # Make sure we have a Tubes channel - PS doesn't yet provide one
        if tubes_chan is None:
            _logger.debug("Didn't find our Tubes channel, requesting one...")
            tubes_chan = self.conn.request_channel(telepathy.CHANNEL_TYPE_TUBES, 
                                                   telepathy.HANDLE_TYPE_ROOM, room, True)

        self.tubes_chan = tubes_chan
        self.text_chan = text_chan

        tubes_chan[telepathy.CHANNEL_TYPE_TUBES].connect_to_signal('NewTube', 
                                                                   self._new_tube_cb)

    def _list_tubes_reply_cb(self, tubes):
        for tube_info in tubes:
            self._new_tube_cb(*tube_info)

    def _list_tubes_error_cb(self, e):
        _logger.debug('ListTubes() failed: %s'%e)

    def _joined_cb(self, activity):
        if not self._shared_activity:
            return

        _logger.debug('Joined an existing shared activity')
        
        self.initiating = False
        self._setup()
                
        _logger.debug('This is not my activity: waiting for a tube...')
        self.tubes_chan[telepathy.CHANNEL_TYPE_TUBES].ListTubes(
            reply_handler=self._list_tubes_reply_cb, 
            error_handler=self._list_tubes_error_cb)

    def _new_tube_cb(self, id, initiator, type, service, params, state):
        _logger.debug('New tube: ID=%d initator=%d type=%d service=%s '
                     'params=%r state=%d' %(id, initiator, type, service, 
                     params, state))

        if (type == telepathy.TUBE_TYPE_DBUS and
            service == SERVICE):
            if state == telepathy.TUBE_STATE_LOCAL_PENDING:
                self.tubes_chan[telepathy.CHANNEL_TYPE_TUBES].AcceptDBusTube(id)

            self.tube_conn = TubeConnection(self.conn, 
                self.tubes_chan[telepathy.CHANNEL_TYPE_TUBES], 
                id, group_iface=self.text_chan[telepathy.CHANNEL_INTERFACE_GROUP])
            
            _logger.debug('Tube created')
            self.messenger = Messenger(self.tube_conn, self.initiating, self.model)         

             
    def _load_homepage(self):
        if os.path.isfile(_LIBRARY_PATH):
            self._browser.load_uri('file://' + _LIBRARY_PATH)
        else:
            self._browser.load_uri('about:blank')

    def _session_history_changed_cb(self, session_history, link):
        _logger.debug('NewPage: %s.' %link)
        self.current = link
        
    def _title_changed_cb(self, embed, pspec):
        if embed.props.title is not '':
            _logger.debug('Title changed=%s' % embed.props.title)
            self.webtitle = embed.props.title
            
    def read_file(self, file_path):
        if self.metadata['mime_type'] == 'text/plain':
            f = open(file_path, 'r')
            try:
                data = f.read()
            finally:
                f.close()
            self.model.deserialize(data)
            
            for link in self.model.data['shared_links']:
                _logger.debug('read: url=%s title=%s d=%s' % (link['url'],
                                                              link['title'],
                                                              link['color']))
                self._add_link_totray(link['url'],
                                      base64.b64decode(link['thumb']),
                                      link['color'], link['title'],
                                      link['owner'], -1, link['hash'])                                
            self._browser.set_session(self.model.data['history'])
        else:
            self._browser.load_uri(file_path)
        
    def write_file(self, file_path):
        if not self.metadata['mime_type']:
            self.metadata['mime_type'] = 'text/plain'
        
        if self.metadata['mime_type'] == 'text/plain':
            if not self._jobject.metadata['title_set_by_user'] == '1':
                if self._browser.props.title:
                    self.metadata['title'] = self._browser.props.title

            self.model.data['history'] = self._browser.get_session()

            f = open(file_path, 'w')
            try:
                f.write(self.model.serialize())
            finally:
                f.close()

    def _link_add_button_cb(self, button):
        _logger.debug('button: Add link: %s.' % self.current)                
        self._add_link()
            
    def _key_press_cb(self, widget, event):
        if event.state & gtk.gdk.CONTROL_MASK:
            if gtk.gdk.keyval_name(event.keyval) == "l":
                _logger.debug('keyboard: Add link: %s.' % self.current)                
                self._add_link()                
                return True
            elif gtk.gdk.keyval_name(event.keyval) == "u":
                _logger.debug('keyboard: Show source of the current page')
                self._show_source()
                return True
            elif gtk.gdk.keyval_name(event.keyval) == "minus":
                _logger.debug('keyboard: Zoom out')
                self._browser.zoom_out()
                return True
            elif gtk.gdk.keyval_name(event.keyval) == "plus" \
                     or gtk.gdk.keyval_name(event.keyval) == "equal" :
                _logger.debug('keyboard: Zoom in')
                self._browser.zoom_in()
                return True
        elif event.keyval == _VIEW_SOURCE_KEY_CODE:
            _logger.debug('keyboard: Show source of the current page SHOW_KEY')
            self._show_source()
            return True
        return False

    def _add_link(self):
        ''' take screenshot and add link info to the model '''
        for link in self.model.data['shared_links']:
            if link['hash'] == sha.new(self.current).hexdigest():
                _logger.debug('_add_link: link exist already a=%s b=%s' %(
                    link['hash'], sha.new(self.current).hexdigest()))
                return
        buffer = self._get_screenshot()
        timestamp = time.time()
        self.model.add_link(self.current, self.webtitle, buffer,
                            profile.get_nick_name(),
                            profile.get_color().to_string(), timestamp)

        if self.messenger is not None:
            self.messenger._add_link(self.current, self.webtitle,                                     
                                     profile.get_color().to_string(),
                                     profile.get_nick_name(),
                                     base64.b64encode(buffer), timestamp)

    def _add_link_model_cb(self, model, index):
        ''' receive index of new link from the model '''
        link = self.model.data['shared_links'][index]
        self._add_link_totray(link['url'], base64.b64decode(link['thumb']),
                              link['color'], link['title'],
                              link['owner'], index, link['hash'])

    def _add_link_totray(self, url, buffer, color, title, owner, index, hash):
        ''' add a link to the tray '''
        item = LinkButton(url, buffer, color, title, owner, index, hash)
        item.connect('clicked', self._link_clicked_cb, url)
        item.connect('remove_link', self._link_removed_cb)
        self._tray.add_item(item, index) # use index to add to the tray
        item.show()
        if self._tray.props.visible is False:
            self._tray.show()        
        self.viewtoolbar.traybutton.props.sensitive = True
        
    def _link_removed_cb(self, button, hash):
        ''' remove a link from tray and delete it in the model '''
        self.model.remove_link(hash)
        self._tray.remove_item(button)
        if len(self._tray.get_children()) == 0:
            self.viewtoolbar.traybutton.props.sensitive = False

    def _link_clicked_cb(self, button, url):
        ''' an item of the link tray has been clicked '''
        self._browser.load_uri(url)
                
    def _show_source(self):
        self._browser.get_source()

    def _pixbuf_save_cb(self, buf, data):
        data[0] += buf
        return True

    def get_buffer(self, pixbuf):
        data = [""]
        pixbuf.save_to_callback(self._pixbuf_save_cb, "png", {}, data)
        return str(data[0])

    def _get_screenshot(self):
        window = self._browser.window
        width, height = window.get_size()

        screenshot = gtk.gdk.Pixbuf(gtk.gdk.COLORSPACE_RGB, has_alpha=False,
                                    bits_per_sample=8, width=width,
                                    height=height)
        screenshot.get_from_drawable(window, window.get_colormap(), 0, 0, 0, 0,
                                     width, height)

        screenshot = screenshot.scale_simple(style.zoom(100),
                                                 style.zoom(80),
                                                 gtk.gdk.INTERP_BILINEAR)

        buffer = self.get_buffer(screenshot)
        return buffer

    def close(self):
        if downloadmanager.can_quit():
            activity.Activity.close(self)        
        else:
            logging.debug('Close is called')
            alert = Alert()
            alert.props.title = _('Download in progress')
            alert.props.msg = _('Stopping now will cancel your download')
            cancel_icon = Icon(icon_name='dialog-cancel')
            alert.add_button(gtk.RESPONSE_CANCEL, _('Cancel'), cancel_icon)
            stop_icon = Icon(icon_name='dialog-ok')
            alert.add_button(gtk.RESPONSE_OK, _('Stop'), stop_icon)
            stop_icon.show()
            self.add_alert(alert)
            alert.connect('response', self.__inprogress_response_cb)
            alert.show()            
            self.present()
            
    def __inprogress_response_cb(self, alert, response_id):
        self.remove_alert(alert)
        if response_id is gtk.RESPONSE_CANCEL:
            logging.debug('Keep on')
        elif response_id == gtk.RESPONSE_OK:
            logging.debug('Stop downloads and quit')
            downloadmanager.remove_all_downloads()
            activity.Activity.close(self)        
            
